src/enclave/enclavetypes.h

```diff
@@ -11,69 +11,124 @@ namespace enclave
-    RPCContext(
-      size_t client_session_id_,
-      const std::vector<uint8_t>& caller_cert_,
-      ccf::ActorsType actor_ = ccf::ActorsType::unknown) :
+    SessionContext(
+      size_t client_session_id_, const std::vector<uint8_t>& caller_cert_) :
       client_session_id(client_session_id_),
-      caller_cert(caller_cert_),
-      actor(actor_)
+      caller_cert(caller_cert_)
     {}
 
     // Constructor used for forwarded and PBFT RPC
-    RPCContext(
+    SessionContext(
       size_t fwd_session_id_,
       ccf::CallerId caller_id_,
-      ccf::ActorsType actor_ = ccf::ActorsType::unknown,
       const std::vector<uint8_t>& caller_cert_ = {}) :
-      fwd(std::make_optional<struct forwarded>(fwd_session_id_, caller_id_)),
-      actor(actor_),
+      fwd(std::make_optional<Forwarded>(fwd_session_id_, caller_id_)),
       caller_cert(caller_cert_)
     {}
   };
 
+  struct RPCContext
+  {
+    SessionContext session;
+
+    // Packing format of original request, should be used to pack response
+    std::optional<jsonrpc::Pack> pack = jsonrpc::Pack::Text;
+
+    // TODO: Avoid unnecessary copies
+    std::vector<uint8_t> raw = {};
+
+    nlohmann::json unpacked_rpc = {};
+
+    std::optional<ccf::SignedReq> signed_request = std::nullopt;
+
+    // Actor type to dispatch to appropriate frontend
+    ccf::ActorsType actor = ccf::ActorsType::unknown;
+
+    // Method indicates specific handler for this request
+    std::string method = {};
+
+    uint64_t seq_no = {};
+
+    nlohmann::json params = {};
+
+    bool is_create_request = false;
+
+    RPCContext(const SessionContext& s) : session(s) {}
+  };
+
+  inline void parse_rpc_context(RPCContext& rpc_ctx, const nlohmann::json& rpc)
+  {
+    const auto sig_it = rpc.find(jsonrpc::SIG);
+    if (sig_it != rpc.end())
+    {
+      rpc_ctx.unpacked_rpc = rpc.at(jsonrpc::REQ);
+      rpc_ctx.signed_request = rpc;
+    }
+    else
+    {
+      rpc_ctx.unpacked_rpc = rpc;
+      rpc_ctx.signed_request = std::nullopt;
+    }
+
+    const auto method_it = rpc_ctx.unpacked_rpc.find(jsonrpc::METHOD);
+    if (method_it != rpc_ctx.unpacked_rpc.end())
+    {
+      rpc_ctx.method = method_it->get<std::string>();
+    }
+
+    const auto seq_it = rpc_ctx.unpacked_rpc.find(jsonrpc::ID);
+    if (seq_it != rpc_ctx.unpacked_rpc.end())
+    {
+      rpc_ctx.seq_no = seq_it->get<uint64_t>();
+    }
+
+    const auto params_it = rpc_ctx.unpacked_rpc.find(jsonrpc::PARAMS);
+    if (params_it != rpc_ctx.unpacked_rpc.end())
+    {
+      rpc_ctx.params = *params_it;
+    }
+  }
+
+  inline RPCContext make_rpc_context(
+    const SessionContext& s, const std::vector<uint8_t>& packed)
+  {
+    RPCContext rpc_ctx(s);
+
+    rpc_ctx.raw = packed;
+
+    auto [success, rpc] = jsonrpc::unpack_rpc(packed, rpc_ctx.pack);
+    if (!success)
+    {
+      throw std::logic_error(fmt::format("Failed to unpack: {}", rpc.dump()));
+    }
+
+    parse_rpc_context(rpc_ctx, rpc);
```

Comment:
What about first un-packing the RPC, then calling `parse_rpc_context` which can create and return a new `RPCContext` object instead of having a `RPCContext` passed into the method and modifying it, and then setting `raw` and `SessionContext` to the returned `RPCContext`?